mono/metadata/threads.c

```diff
@@ -943,19 +943,26 @@ mono_thread_detach_internal (MonoInternalThread *thread)
 	thread->current_appcontext = NULL;
 
 	/*
+	 * thread->synch_cs can be NULL if this was called after
+	 * ves_icall_System_Threading_InternalThread_Thread_free_internal.
+	 * This can happen only during shutdown.
+	 * The shutting_down flag is not always set, so we can't assert on it.
 	 * This should be alive until after the reference queue runs the
 	 * post-free cleanup function
+	 *
+	 * The lock here also helps us synchronize with
+	 * mono_threads_suspend_all_other_threads.  We don't want to delete
+	 * thread->suspended if suspend_all_other_threads is holding the thread
+	 * lock.
 	 */
-	while (TRUE) {
-		guint32 old_state = thread->state;
+	if (thread->synch_cs)
+		LOCK_THREAD (thread);
 
-		guint32 new_state = old_state;
-		new_state |= ThreadState_Stopped;
-		new_state &= ~ThreadState_Background;
+	thread->state |= ThreadState_Stopped;
+	thread->state &= ~ThreadState_Background;
 
-		if (mono_atomic_cas_i32 ((gint32 *)&thread->state, new_state, old_state) == old_state)
-			break;
-	}
+	if (thread->synch_cs)
+		UNLOCK_THREAD (thread);
```

Comment:
I still like to try to guarantee the cs stays valid until the unlock. By refcounting it. Or at least storing it in local. Granted, might already be otherwise guaranteed. You know, is second if redundant? Can it be false but first true? Or vice versa? That’d be pretty bad, if possible.

Reply:
Also, if no cs, then interlock? What does happen if the cs not there?

Reply:
I believe with the refqueue freeing of synch_cs, there is only one situation where we get here with `thread->synch_cs == NULL`  --- when we're detaching the finalizer thread.  In which case nothing else is running, so it's probably alright to change the finalizer thread's state without locking.  But yea these lifetimes are not great.